src/jaegertracing/Tracer.cpp

```diff
@@ -67,19 +70,32 @@ Tracer::StartSpanWithOptions(string_view operationName,
     try {
         const auto result = analyzeReferences(options.references);
         const auto* parent = result._parent;
+        const auto* self = result._self;
         const auto& references = result._references;
+        if (self && (parent || !references.empty()))
```

Comment:
When looking at the Go implementation, having a parent/references does not cause an error. They are simply ignored.

Is there any reason to do it differently here?

Reply:
1)
About unit tests. I tried to add a UT a few times and it always ended with CI failing with an Access Denied during calling of StartSpan with ChildOf.
I'm 99.9% sure the reason is in dependencies (in opentracing.lib/.dll to be more precise) because I had very much the same problem when I tried to run the App.cpp on my Win10+VS 2019 computer.
I fixed it locally by building opentracing-cpp 1.5.x project and using its results in jaegertracing code.

To prove that my changes have nothing to do with the Access Denied, I created another temp PR where the only change is in TracerTest.cpp and received the same error: https://github.com/jaegertracing/jaeger-client-cpp/pull/208
I'm not familiar with Hunter, so cannot fix this problem myself.

That's the reason I couldn't add any changes in unittests.

----

2) 
About returning nullptr if SelfRef is used: I thought that if a library user is making a mistake of setting both parent span and selfref, the library should react with some type of mistake (in this case it would return a nullptr). IMO creating a root span instead of a requested child span in this case is closer to "unspecified behavior" than a reaction to an error.
If you think the latter behavior is better for the project, I'll reimplement it.

Reply:
I will check your test tomorrow as I ll have access to a Windows machine and will try to help you on that.

Reply:
Hi @Alek86, I tried your change locally on my Windows machine and indeed, stumbled on the same error.

I then updated the OpenTracing version to 1.6.0 and the exception was fixed. In  fact, there is a problem with the debug flags of OpenTracing that was fixed with the latest release. That bug was leading to different object layout between OpenTracing lib and Jaeger one, leading to that behavior.

I will start by updating OpenTracing dependency to 1.6.0.

Reply:
Thank you! I suppose I should wait for your fix (in master?), then merge it to this branch and then add the UT.
Please also let me know if you'll accept returning nullptr behavior or still want to ignore ChildOf etc if SelfRef is passed

Reply:
hello @Alek86, I opened the PR #209 to upgrade to OpenTracing 1.6.0. This should solve your crash under WIndows.

Reply:
Hello @Alek86, PR #209  was merged, You can rebase now and check the unit test.

Reply:
Hi @mdouaihy, I've added the unit tests

Reply:
Hi @Alek86, thank you for your effort.

One more think I just noticed, IMO, I think that if we throw an exception when the span state is invalid, maybe we can throw a typed exception with the relevant information.   std::exception() is too generic. What do you think?

Reply:
Okay. Do you have a place in mind where I should add the new exception class?

Reply:
Replaced std::exception with std::invalid_argument with explanation text
Added the explanation text to the error log

Reply:
I believe that the message should state that the Self reference and other types of references are exclusive. The current message informs only about self. What do you think?

Reply:
I think that "self reference must be the only reference" text also means that the other types cannot be used with self ref.
If you prefer to see another text, can you give it to me, I'll use it?

Reply:
I suggest: Self and references are exclusive. Only one can be specified. What do you think?

Reply:
Other references do not need to come as "only one".

Suggestion: "Self reference cannot be combined with other reference types"

Reply:
I'm ready for any text change that will help this code reach master :)
I'll add "Self and references are exclusive. Only one of them can be specified"

Reply:
Done